### PR TITLE
[#1152] fix: Direct memory may leak in shuffle server.

### PR DESCRIPTION
**Pull request by zhengchenyu**
### What changes were proposed in this pull request?

Release buffer when since the buffer have copied to data of reply.

### Why are the changes needed?

Fix: #1152

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Test in cluster manually.

### CODE REVIEW

**Comment by jerqi** [comment]
@leixm Could you help me review this pr?

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java [hunk 1]
```diff
@@ -658,6 +658,9 @@ public void getLocalShuffleData(
                 .setRetMsg(msg)
                 .build();
       } finally {
+        if (sdr != null) {
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java:661]: Maybe we should release `ShuffleIndexResult`, too.

Comment (zhengchenyu) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java:661]: @jerqi 
In fact, getLocalShuffleData and getLocalShuffleIndex only use FileSegmentManagedBuffer. It means will trigger direct  memory leak. I fix this only to keep pace with getMemoryShuffleData.
Do you think we should fix ShuffleIndexResult to keep pace with this? Or remove the fix on getLocalShuffleData?

Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java:661]: Yes, we can remove the fix on `getLocalShuffleData` and add some comments to explain why we don't need the fix. Or we fix the method `getLocalShuffleIndex`, too.

**Review by zuston — COMMENTED** [review]
If so, memory leaks are a must-have problem. Right?

**Comment by LuciferYang** [comment]
Is this the only case? Maybe we should run some load tests and use some tools to analyze off-heap memory usage, to make this bug fix as thorough as possible?

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java (hunk 1 above)
Comment (zhengchenyu) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java:661]: > Yes, we can remove the fix on `getLocalShuffleData` and add some comments to explain why we don't need the fix. Or we fix the method `getLocalShuffleIndex`, too.

Ok, I will fix the method `getLocalShuffleIndex`.
For now, even though FileSegmentManagedBuffer::release do nothing,  I am afraid that allocate some resource in the future.

**Comment by zhengchenyu** [comment]
> If so, memory leaks are a must-have problem. Right?

For now, I think it is deterministic problem if client type is GRPC but not GRPC_NETTY.

**Comment by zhengchenyu** [comment]
> Is this the only case? Maybe we should run some load tests and use some tools to analyze off-heap memory usage, to make this bug fix as thorough as possible?

Yes, @qijiale76 have create issue #1153 to discuss this problem.

**Comment by zuston** [comment]
> > If so, memory leaks are a must-have problem. Right?
> 
> For now, I think it is deterministic problem if client type is GRPC but not GRPC_NETTY.

Got it, it looks the bug was introduced by netty implementation.

**Comment by LuciferYang** [comment]
> > > If so, memory leaks are a must-have problem. Right?
> > 
> > 
> > For now, I think it is deterministic problem if client type is GRPC but not GRPC_NETTY.
> 
> Got it, it looks the bug was introduced by netty implementation.

IIRC, Netty has a mechanism to detect memory leaks within itself. Is it that this feature is not enabled, or is the leaked memory something Netty is unable to detect?

**Review by jerqi — APPROVED** [review]
LGTM

**Comment by jerqi** [comment]
> > > > If so, memory leaks are a must-have problem. Right?
> > > 
> > > 
> > > For now, I think it is deterministic problem if client type is GRPC but not GRPC_NETTY.
> > 
> > 
> > Got it, it looks the bug was introduced by netty implementation.
> 
> IIRC, Netty has a mechanism to detect memory leaks within itself. Is it that this feature is not enabled, or is the leaked memory something Netty is unable to detect?

This is found by zhengchenyu with the option `-Dio.netty.leakDetection.level=PARANOID -XX:NativeMemoryTracking=detail`

**Comment by LuciferYang** [comment]
> > > > > If so, memory leaks are a must-have problem. Right?
> > > > 
> > > > 
> > > > For now, I think it is deterministic problem if client type is GRPC but not GRPC_NETTY.
> > > 
> > > 
> > > Got it, it looks the bug was introduced by netty implementation.
> > 
> > 
> > IIRC, Netty has a mechanism to detect memory leaks within itself. Is it that this feature is not enabled, or is the leaked memory something Netty is unable to detect?
> 
> This bug is found by zhengchenyu with the option `-Dio.netty.leakDetection.level=PARANOID -XX:NativeMemoryTracking=detail`

Got it

**Comment by zhengchenyu** [comment]
截屏2023-08-17 下午2 51 36

Before this PR, direct memory grows rapidly to limited 80G, just see blue.
After this PR, is the green line.

**Comment by jerqi** [comment]
Merged to master.

**Pull request merged**